File: routes/voice.py
import os
from datetime import datetime

from fastapi import APIRouter, HTTPException, UploadFile, File, Form
from fastapi.responses import FileResponse

from database import get_db

router = APIRouter(tags=["voice"])


@router.get("/api/voice")
def list_voice_notes(question_id: int):
    conn = get_db()
    rows = [dict(r) for r in conn.execute(
        "SELECT * FROM voice_notes WHERE question_id = ? ORDER BY created_at DESC",
        (question_id,),
    )]
    conn.close()
    return rows


@router.post("/api/voice/{question_id}", status_code=201)
async def create_voice_note(
    question_id: int,
    file: UploadFile = File(...),
    duration: float = Form(0),
    title: str = Form(""),
):
    content = await file.read()
    ext = (file.filename or "recording.webm").rsplit(".", 1)[-1]
    ts = datetime.now().strftime("%Y%m%d_%H%M%S")
    filename = f"q{question_id}_{ts}.{ext}"
    path = os.path.join("recordings", filename)
    with open(path, "wb") as f:
        f.write(content)

    conn = get_db()
    cur = conn.execute(
        "INSERT INTO voice_notes (question_id, filename, duration_seconds, title) VALUES (?, ?, ?, ?)",
        (question_id, filename, duration, title),
    )
    vid = cur.lastrowid
    conn.commit()
    row = dict(conn.execute("SELECT * FROM voice_notes WHERE id = ?", (vid,)).fetchone())
    conn.close()
    return row


@router.delete("/api/voice/{note_id}")
def delete_voice_note(note_id: int):
    conn = get_db()
    row = conn.execute("SELECT filename FROM voice_notes WHERE id = ?", (note_id,)).fetchone()
    if row:
        path = os.path.join("recordings", row["filename"])
        if os.path.exists(path):
            os.remove(path)
    conn.execute("DELETE FROM voice_notes WHERE id = ?", (note_id,))
    conn.commit()
    conn.close()
    return {"ok": True}


@router.get("/recordings/{filename}")
def get_recording(filename: str):
    path = os.path.join("recordings", filename)
    if not os.path.exists(path):
        raise HTTPException(404, "Recording not found")
    return FileResponse(path)

File: server/routes/voice.js
const express = require('express')
const router = express.Router()
const multer = require('multer')
const path = require('path')
const fs = require('fs')
const { getDb } = require('../db')
const { RECORDINGS_DIR } = require('../paths')
const { v4: uuidv4 } = require('uuid')

if (!fs.existsSync(RECORDINGS_DIR)) fs.mkdirSync(RECORDINGS_DIR, { recursive: true })

const storage = multer.diskStorage({
  destination: RECORDINGS_DIR,
  filename: (req, file, cb) => cb(null, `${uuidv4()}.webm`)
})
const upload = multer({ storage, limits: { fileSize: 50 * 1024 * 1024 } })

router.get('/', (req, res) => {
  const db = getDb()
  const { file_id } = req.query
  const rows = file_id
    ? db.prepare('SELECT * FROM voice_notes WHERE file_id = ? ORDER BY created_at DESC').all(file_id)
    : db.prepare('SELECT * FROM voice_notes ORDER BY created_at DESC').all()
  res.json(rows)
})

router.post('/', upload.single('audio'), (req, res) => {
  const db = getDb()
  const { file_id, duration, title } = req.body
  if (!req.file) return res.status(400).json({ error: 'No audio file' })

  const id = uuidv4()
  db.prepare(
    "INSERT INTO voice_notes (id, file_id, filename, duration_seconds, title, created_at) VALUES (?, ?, ?, ?, ?, datetime('now'))"
  ).run(id, file_id || null, req.file.filename, parseInt(duration) || 0, title || null)
  res.status(201).json(db.prepare('SELECT * FROM voice_notes WHERE id = ?').get(id))
})

router.delete('/:id', (req, res) => {
  const db = getDb()
  const note = db.prepare('SELECT * FROM voice_notes WHERE id = ?').get(req.params.id)
  if (!note) return res.status(404).json({ error: 'Not found' })
  try { fs.unlinkSync(path.join(RECORDINGS_DIR, note.filename)) } catch (_) {}
  db.prepare('DELETE FROM voice_notes WHERE id = ?').run(req.params.id)
  res.json({ ok: true })
})

module.exports = router
